axelor-base/src/main/java/com/axelor/apps/base/service/RegistrationNumberValidationFRA.java

```diff
@@ -0,0 +1,154 @@
+    int siren = Integer.parseInt(sirenStr);
+    int taxKey = Math.floorMod(siren, 97);
+    taxKey = Math.floorMod(12 + 3 * taxKey, 97);
+    return String.format("%02d", taxKey);
+  }
+
+  protected String getTaxNbrFromRegistrationCode(Partner partner) {
+    String taxNbr = "";
+
+    if (partner.getMainAddress() != null
+        && partner.getMainAddress().getAddressL7Country() != null) {
+      String countryCode = partner.getMainAddress().getAddressL7Country().getAlpha2Code();
+      String regCode = partner.getRegistrationCode();
+
+      if (regCode != null) {
+        regCode = regCode.replaceAll(" ", "");
+
+        if (regCode.length() == 14) {
+          String siren = regCode.substring(0, 9);
+          String taxKey = getTaxKeyFromSIREN(siren);
+
+          taxNbr = String.format("%s%s%s", countryCode, taxKey, siren);
+        }
+      }
+    }
+
+    return taxNbr;
+  }
+
+  protected String getNicFromRegistrationCode(Partner partner) {
+    String regCode = partner.getRegistrationCode();
+    String nic = "";
+
+    if (partner.getMainAddress() != null
+        && partner.getMainAddress().getAddressL7Country() != null) {
+      Country businessCountry = partner.getMainAddress().getAddressL7Country();
+      RegistrationNumberTemplate registrationNumberTemplate =
+          businessCountry.getRegistrationNumberTemplate();
+      if (registrationNumberTemplate.getUseNic() && regCode != null) {
+        regCode = regCode.replaceAll(" ", "");
+
+        if (regCode.length() == registrationNumberTemplate.getRequiredSize()) {
+          nic =
+              regCode.substring(
+                  registrationNumberTemplate.getNicPos() - 1,
+                  registrationNumberTemplate.getNicPos()
+                      + registrationNumberTemplate.getNicLength()
+                      - 1);
+        }
+      }
+    }
+
+    return nic;
+  }
+
+  public String getSirenFromRegistrationCode(Partner partner) {
+    String regCode = partner.getRegistrationCode();
+    String siren = "";
+
+    if (partner.getMainAddress() != null
+        && partner.getMainAddress().getAddressL7Country() != null) {
+      Country businessCountry = partner.getMainAddress().getAddressL7Country();
+      RegistrationNumberTemplate registrationNumberTemplate =
+          businessCountry.getRegistrationNumberTemplate();
+      if (registrationNumberTemplate.getUseSiren() && regCode != null) {
+        regCode = regCode.replaceAll(" ", "");
+
+        if (regCode.length() == registrationNumberTemplate.getRequiredSize()) {
+          siren =
+              regCode.substring(
+                  registrationNumberTemplate.getSirenPos() - 1,
+                  registrationNumberTemplate.getSirenPos()
+                      + registrationNumberTemplate.getSirenLength()
+                      - 1);
+        }
+      }
+    }
+    return siren;
+  }
+
+  @Override
+  public Map<String, Map<String, Object>> getRegistrationCodeValidationAttrs(Partner partner) {
```

Comment:
this method should be in RegistrationNumberValidation, it's a generic method to all the Validation methods.